### PR TITLE
Support infinity loading A-Z/Z-A, set raduis ratio in xml attr.

### PR DESCRIPTION
**Pull request by KhaledLela**
Supporting infinity item displaying > Will be helpful when display half wheel.
Support Load more from server with pagination.
Support Load more with sort= {A-Z , Z-A}.

### CODE REVIEW

**Comment by szugyi** [comment]
Hi @KhaledLela!

Thanks for contributing to the library!

I wanted to run the app to see the improvements you made, but the sample app could not even compile as there are quite a few breaking changes. I would not like to change the library that drastically all at once, but lets discuss it later. Could you please update the sample in a way, that the new functionality of the WheelBase class can be tested?

**Comment by KhaledLela** [comment]
@szugyi Samples updated, and i added Spin wheel sample,

**Comment by szugyi** [comment]
Dear @KhaledLela!

Thank you for the sample.
The concept of the wheel is pretty good, and usage like this has been requested before. However I have to be honest, I do not see the practicality in infinite loading on a circle. I think it is confusing. Would you mind to just add a "Wheel of fortune" functionality? 

I would not be able to maintain the added complexity either. You could release this enhancement as a separate library I think, since it is quite different from the original intention.

**Comment by KhaledLela** [comment]
Dear @szugyi 
Thanks for fast response, 
> I do not see the practicality in infinite loading on a circle. I think it is confusing.
I am agree with your opinion, circular menu is to display **finite items**.
But this was a feature inside our company apps, To load sub categories (n items by pagination) for each main category.
 - When search about infinity items wheel, I see that it required feature.
> Would you mind to just add a "Wheel of fortune" functionality?
 - Great, To add to your fantastic library.

- Supporting infinity items it's up to you,
Thanks for your great effort,

File: circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+            onRotateListener.onRotationFinished(view);
+    }
+
+    public interface WheelItem{
+        String getName();
+        long getId();
+    }
+
+    public interface OnRotateListener extends OnRotationFinishedListener{
+        void onSettleRotation(float endDegree, long duration);
```
Comment (szugyi) [inline circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java]: settle? more like start I guess

File: circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+        void onRotate(float degree);
+        void onStopAnimation();
```
Comment (szugyi) [inline circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java]: why not onRotationFinished?

File: circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    }
+
+    public interface OnLoadMoreListener {
+        void loadMore(boolean isNext);
```
Comment (szugyi) [inline circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java]: direction should be indicated more clearly

File: circlemenu/src/main/res/values/attrs.xml [hunk 4]
```diff
@@ -18,6 +18,8 @@
         <attr name="isRotating" format="boolean" />
         <attr name="radius" format="dimension" />
         <attr name="speed" format="integer" />
+        <attr name="wheel_radius_ratio" format="integer" />
```
Comment (szugyi) [inline circlemenu/src/main/res/values/attrs.xml]: why does it has to be an integer? could be float

File: sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    private WheelLayout scWheel;
+    private final int ITEM_LIMIT = 20;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_wheel);
+        final View wheelView = findViewById(R.id.wheel_layout);
+        mcWheel = (WheelLayout) findViewById(R.id.mc_wheel);
+        scWheel = (WheelLayout) findViewById(R.id.sc_wheel);
```
Comment (szugyi) [inline sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java]: mc? sc?

File: sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        });
+        initWheel();
+        loadMC();
+        loadSCByMC(mcWheel.getSelectedWheelItem());
+    }
+
+    private void initWheel() {
+        mcWheel.setBgImageView((ImageView) findViewById(R.id.mc_bg));
+        scWheel.setBgImageView((ImageView) findViewById(R.id.sc_bg));
+        final ImageView sprLeft = (ImageView) findViewById(R.id.spr_left);
```
Comment (szugyi) [inline sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java]: spr?

File: sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                if (animator != null && animator.isRunning()) {
+                    animator.cancel();
+                    animator = null;
+                }
+            }
+
+            @Override
+            public void onRotate(float angle) {
+                final float rotation = sprRight.getRotation() - angle;
+                if (rotation > 83 || rotation <-23)
```
Comment (szugyi) [inline sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java]: what are these magic numbers?

File: sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    private void loadSCByMC(WheelLayout.WheelItem mc,boolean isNext) {
+        final int itemCount = scWheel.getItemCount();
+        final int start = isNext ? itemCount : itemCount + ITEM_LIMIT;
+        final int end = isNext ? itemCount + ITEM_LIMIT : itemCount;
+        loadSCByMC(mc, start, end);
+    }
+
+    private void loadSCByMC(WheelLayout.WheelItem mc, int start, int end) {
+        List<MenuItem> items = new ArrayList<>(ITEM_LIMIT);
+        for (int i=start ; i<end;i++){
```
Comment (szugyi) [inline sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java]: Formatting

File: sample/src/main/res/layout/activity_wheel.xml [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+            <com.szugyi.circlemenu.view.WheelLayout
+                android:id="@+id/mc_wheel"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_gravity="center"
+                app:speed="50"
+                app:wheel_radius_ratio="@dimen/spin_radius_ratio"
+                app:firstChildPosition="East">
+                <include layout="@layout/wheel_items" />
```
Comment (szugyi) [inline sample/src/main/res/layout/activity_wheel.xml]: This include is almost invisible. Put an enter before it please

File: sample/src/main/res/layout/activity_wheel.xml [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            android:id="@+id/spin_spr_view"
+            android:layout_gravity="center"
+            android:layout_width="@dimen/spin_pointer_width"
+            android:layout_height="@dimen/spin_pointer_width">
+            <ImageView
+                android:id="@+id/spr_left"
+                android:layout_width="wrap_content"
+                android:src="@drawable/spin_wheel_spr_blue"
+                android:layout_gravity="left|center_vertical"
+                android:rotation="-30"
```
Comment (szugyi) [inline sample/src/main/res/layout/activity_wheel.xml]: You would not need to rotate the image immediately if it was not a rotated image in the first place. Why is the `spin_wheel_spr_blue` drawable rotated and not completely horizontal?

File: sample/src/main/res/layout/wheel_items.xml [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+    <TextView
+        android:id="@+id/item_2"
+        android:layout_width="@dimen/spin_item_width"
+        android:layout_height="wrap_content"
+        android:lines="2"
+        android:textColor="@android:color/white"
+        android:gravity="center"
+        android:textSize="@dimen/txt_h4"
+        android:rotation="45"/>
```
Comment (szugyi) [inline sample/src/main/res/layout/wheel_items.xml]: It should not be needed to rotate the children in advance

File: sample/src/main/res/values-large-mdpi/dimens.xml [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    <item name="spin_radius_ratio" type="dimen" format="float">4.5</item>
+    <dimen name="spin_pointer_width">120dp</dimen>
+
+    <dimen name="spin_pointer_pivot_x_left">60dp</dimen>
+    <dimen name="spin_pointer_pivot_y_left">40dp</dimen>
+
+    <dimen name="spin_pointer_pivot_x_right">20dp</dimen>
+    <dimen name="spin_pointer_pivot_y_right">40dp</dimen>
+
+    <!-- 1. Fonts sizes-->
```
Comment (szugyi) [inline sample/src/main/res/values-large-mdpi/dimens.xml]: ?

**Review by szugyi — CHANGES_REQUESTED** [review]

**Comment by szugyi** [comment]
Hi @KhaledLela!

I just want to let you know, that I will make the library deprecated, as the new version of ConstraintLayout can basically do the same thing: https://medium.com/devnibbles/constraintlayout-circular-positioning-9489b11cb0e5

If you would like to publish your version using this library as the base feel free to do so.

File: circlemenu/src/main/java/com/szugyi/circlemenu/view/CircleLayout.java [hunk 13]
```diff
@@ -98,52 +102,57 @@ public CircleLayout(Context context, AttributeSet attrs) {
...
+        Log.d("WheelMenu","radiusRatio:" + radiusRatio);
     }
 
     public float getAngle() {
         return angle;
     }
 
-    public void setAngle(float angle) {
-        this.angle = angle % 360;
+    public void setAngle(float mAngle) {
```
[the remark this answers is not in the document]
Comment (KhaledLela) [reply circlemenu/src/main/java/com/szugyi/circlemenu/view/CircleLayout.java]: Ok, thanks for sharing this topic.

File: circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+            itemIndex = mCurrentItem - halfItems;
+            if (itemIndex - BUFFER_COUNT < 0)
+                loadMore(false); // Load more from Z-A
+
+            if (itemIndex <0)
+                itemIndex += mItemCount;
+            viewIndex = selectedIndex + halfItems + 1;
+        }
+
+        final View children = getChildAt(viewIndex); // Get View
```
[the remark this answers is not in the document]
Comment (KhaledLela) [reply circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java]: Yup, You are right 
> children  to child

File: circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+    }
+
+    private void setChildrenValue(View children, WheelItem item) {
+        if (children instanceof TextView) {
+            ((TextView) children).setText(item.getName());
+            children.setTag(R.id.wheel_item_id,item);
+        }
+    }
+
+    public void addItems(List<? extends WheelItem> cats) {
```
[the remark this answers is not in the document]
Comment (KhaledLela) [reply circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java]: its from my implementation cats for categories,
cats > items

File: circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java (hunk 2 above)
Comment (KhaledLela) [reply circlemenu/src/main/java/com/szugyi/circlemenu/view/WheelLayout.java]: onRotationFinished called after animation finished, 
onSettleRotation > may be onAmimationStarted is more descriptive,

File: circlemenu/src/main/res/values/attrs.xml (hunk 4 above)
Comment (KhaledLela) [reply circlemenu/src/main/res/values/attrs.xml]: you are right,

File: sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java (hunk 7 above)
Comment (KhaledLela) [reply sample/src/main/java/com/szugyi/circlemenusample/WheelActivity.java]: Limit for spin pointer rotation for sampling.

**Comment by szugyi** [comment]
Dear @KhaledLela!

Sorry for the long delay.
I still feel like your pull request is really specific to your companys problem, and I would not be able to maintain this complicated functionality. I will close this PR.

If you want to publish your own library using the source of my lib as the base, please feel free to do so. I can refer your library in the readme if you would like.

**Comment by KhaledLela** [comment]
Dear @szugyi 
   Thanks for your time, i really enjoy using your library, also i learned a lot  from following PR guidance,
I don't have enough time right now, but i will do ASAP,
thanks,

**Pull request closed without merging**